src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -140,14 +142,14 @@ public Map<Integer, Float> searchLeaf(LeafReaderContext context, int k) throws I
          * This improves the recall.
          */
         if (isFilteredExactSearchPreferred(cardinality)) {
-            return doExactSearch(context, filterBitSet, k);
+            return doExactSearch(context, new BitSetIterator(filterBitSet, filterBitSet.cardinality()), k);
         }
         Map<Integer, Float> docIdsToScoreMap = doANNSearch(context, filterBitSet, cardinality, k);
         // See whether we have to perform exact search based on approx search results
         // This is required if there are no native engine files or if approximate search returned
         // results less than K, though we have more than k filtered docs
         if (isExactSearchRequire(context, cardinality, docIdsToScoreMap.size())) {
-            final BitSet docs = filterWeight != null ? filterBitSet : null;
+            final BitSetIterator docs = filterWeight != null ? new BitSetIterator(filterBitSet, filterBitSet.cardinality()) : null;
```

Comment:
use the `cardinality` variable retrieved in line 132. Multiple calls to cardinality adds extra latency in the flow.